lib/ukstore/include/uk/store.h

```diff
@@ -0,0 +1,452 @@
+
+#define uk_store_folder_init(folder, f_name)
+
+#define uk_store_add_folder(folder)
+
+#define uk_store_del_folder(folder)
+
+#define uk_store_add_folder_entry(folder, folder_entry)
+
+#define uk_store_del_folder_entry(folder_entry)
+
+#else /* !CONFIG_LIBUKSTORE */
+
+/* Library-specific information generated at compile-time */
+#include <uk/_store_libs.h>
+
+/* Do not call directly */
+#define _uk_store_libid(libname)	\
+	__UK_STORE_ ## libname
+
+/**
+ * Returns the id of a given library.
+ *
+ * @return the id
+ */
+#define uk_store_libid(libname)	\
+	_uk_store_libid(libname)
+
+#endif /* CONFIG_LIBUKSTORE */
+
+/**
+ * Returns the name of the library
+ *
+ * @return this libraries' name
+ */
+#define uk_store_self_name()	\
+	__LIBNAME__
+
+/**
+ * Returns the id of the place it was called from.
+ *
+ * @return this libraries' id
+ */
+#define uk_store_libid_self()	\
+	uk_store_libid(uk_store_self_name())
+
+/* All basic types that exist - use these when interacting with the API */
+enum uk_store_entry_basic_type {
+	none,
+	s8,
+	u8,
+	s16,
+	u16,
+	s32,
+	u32,
+	s64,
+	u64,
+	uptr
+};
+
+/* Getter definitions */
+typedef __s8 (*uk_store_get_s8_func_t)(void);
+typedef __u8 (*uk_store_get_u8_func_t)(void);
+typedef __s16 (*uk_store_get_s16_func_t)(void);
+typedef __u16 (*uk_store_get_u16_func_t)(void);
+typedef __s32 (*uk_store_get_s32_func_t)(void);
+typedef __u32 (*uk_store_get_u32_func_t)(void);
+typedef __s64 (*uk_store_get_s64_func_t)(void);
+typedef __u64 (*uk_store_get_u64_func_t)(void);
+typedef __uptr (*uk_store_get_uptr_func_t)(void);
+
+/* Setter definitions */
+typedef void (*uk_store_set_s8_func_t)(__s8);
+typedef void (*uk_store_set_u8_func_t)(__u8);
+typedef void (*uk_store_set_s16_func_t)(__s16);
+typedef void (*uk_store_set_u16_func_t)(__u16);
+typedef void (*uk_store_set_s32_func_t)(__s32);
+typedef void (*uk_store_set_u32_func_t)(__u32);
+typedef void (*uk_store_set_s64_func_t)(__s64);
+typedef void (*uk_store_set_u64_func_t)(__u64);
+typedef void (*uk_store_set_uptr_func_t)(__uptr);
+
+struct uk_store_entry {
+	/* Function getter pointer */
+	union {
+		uk_store_get_s8_func_t   s8;
+		uk_store_get_u8_func_t   u8;
+		uk_store_get_s16_func_t  s16;
+		uk_store_get_u16_func_t  u16;
+		uk_store_get_s32_func_t  s32;
+		uk_store_get_u32_func_t  u32;
+		uk_store_get_s64_func_t  s64;
+		uk_store_get_u64_func_t  u64;
+		uk_store_get_uptr_func_t uptr;
+	} get;
+
+	/* Function setter pointer */
+	union {
+		uk_store_set_s8_func_t   s8;
+		uk_store_set_u8_func_t   u8;
+		uk_store_set_s16_func_t  s16;
+		uk_store_set_u16_func_t  u16;
+		uk_store_set_s32_func_t  s32;
+		uk_store_set_u32_func_t  u32;
+		uk_store_set_s64_func_t  s64;
+		uk_store_set_u64_func_t  u64;
+		uk_store_set_uptr_func_t uptr;
+	} set;
+
+	/* The entry name */
+	const char *entry_name;
+
+	/* Entry flags */
+	__u16 flags;
+
+	/* Entry getter/setter type */
+	__u16 type;
+} __align8;
+
+struct uk_store_folder {
```

Comment:
Maybe you want to make this header private to the library and define it within the C file. Here you would only do the prototype:
``` c
struct uk_store_folder;
```

I was actually wondering if we should allocate all dynamic libraries on registration within the library with malloc. THis is making the alloc and release scheme easier.

Reply:
There are macros that use fields from `struct uk_store_folder;` inside the header. Same thing for functions inside the `store.c` file. I would keep it like this to avoid complications.

Reply:
Done.